pkg/client/clientcmd/api/types.go

```diff
@@ -0,0 +1,119 @@
+type Config struct {
+	api.TypeMeta `json:",inline"`
+	// Preferences holds general information to be use for cli interactions
+	Preferences Preferences `json:"preferences"`
+	// Clusters is a map of referencable names to cluster configs
+	Clusters map[string]Cluster `json:"clusters"`
+	// AuthInfos is a map of referencable names to user configs
+	AuthInfos map[string]AuthInfo `json:"users"`
+	// Contexts is a map of referencable names to context configs
+	Contexts map[string]Context `json:"contexts"`
+	// CurrentContext is the name of the context that you would like to use by default
+	CurrentContext string `json:"current-context"`
+	// Extensions holds additional information. This is useful for extenders so that reads and writes don't clobber unknown fields
+	Extensions map[string]runtime.EmbeddedObject `json:"extensions,omitempty"`
+}
+
+type Preferences struct {
+	Colors bool `json:"colors,omitempty"`
+	// Extensions holds additional information. This is useful for extenders so that reads and writes don't clobber unknown fields
+	Extensions map[string]runtime.EmbeddedObject `json:"extensions,omitempty"`
+}
+
+// Cluster contains information about how to communicate with a kubernetes cluster
+type Cluster struct {
+	// Server is the address of the kubernetes cluster (https://hostname:port).
+	Server string `json:"server"`
+	// APIVersion is the preferred api version for communicating with the kubernetes cluster (v1beta1, v1beta2, v1beta3, etc).
+	APIVersion string `json:"api-version,omitempty"`
+	// InsecureSkipTLSVerify skips the validity check for the server's certificate. This will make your HTTPS connections insecure.
+	InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
+	// CertificateAuthority is the path to a cert file for the certificate authority.
+	CertificateAuthority string `json:"certificate-authority,omitempty"`
+	// Extensions holds additional information. This is useful for extenders so that reads and writes don't clobber unknown fields
+	Extensions map[string]runtime.EmbeddedObject `json:"extensions,omitempty"`
+}
+
+// AuthInfo contains information that describes identity information.  This is use to tell the kubernetes cluster who you are.
+type AuthInfo struct {
+	// AuthPath is the path to a kubernetes auth file (~/.kubernetes_auth).  If you provide an AuthPath, the other options specified are ignored
+	AuthPath string `json:"auth-path,omitempty"`
+	// ClientCertificate is the path to a client cert file for TLS.
+	ClientCertificate string `json:"client-certificate,omitempty"`
+	// ClientKey is the path to a client key file for TLS.
+	ClientKey string `json:"client-key,omitempty"`
+	// Token is the bearer token for authentication to the kubernetes cluster.
+	Token string `json:"token,omitempty"`
+	// Extensions holds additional information. This is useful for extenders so that reads and writes don't clobber unknown fields
+	Extensions map[string]runtime.EmbeddedObject `json:"extensions,omitempty"`
```

Comment:
The external (versioned) data structure shouldn't be map[string]runtime.EmbeddedObject, it should be:

```
[]struct{
  Name string
  Object runtime.EmbeddedObject
}
```

There were several issues where we agreed that maps in general are not easy for someone to understand the schema of when presented as yaml, and are much harder to describe in an automated fashion to end users.  That's it.

----- Original Message -----
> > +	// Extensions holds additional information. This is useful for extenders
> > so that reads and writes don't clobber unknown fields
> > +	Extensions map[string]runtime.EmbeddedObject
> > `json:"extensions,omitempty"`
> > +}
> > +
> > +// AuthInfo contains information that describes identity information.
> > This is use to tell the kubernetes cluster who you are.
> > +type AuthInfo struct {
> > +	// AuthPath is the path to a kubernetes auth file (~/.kubernetes_auth).
> > If you provide an AuthPath, the other options specified are ignored
> > +	AuthPath string `json:"auth-path,omitempty"`
> > +	// ClientCertificate is the path to a client cert file for TLS.
> > +	ClientCertificate string `json:"client-certificate,omitempty"`
> > +	// ClientKey is the path to a client key file for TLS.
> > +	ClientKey string `json:"client-key,omitempty"`
> > +	// Token is the bearer token for authentication to the kubernetes
> > cluster.
> > +	Token string `json:"token,omitempty"`
> > +	// Extensions holds additional information. This is useful for extenders
> > so that reads and writes don't clobber unknown fields
> > +	Extensions map[string]runtime.EmbeddedObject
> > `json:"extensions,omitempty"`
> 
> > Use a list of keys with values, not maps.
> 
> I'm not quite clear on this.  I want to support an arbitrary number of
> potential extenders that all obey the EmbeddedObject rules.  I want those
> extenders to be able to quickly find their EmbeddedObject which may not be
> unique in type.  Why wouldn't I use a map to express that?  How could an
> extender reliably find their EmbeddedObject (not unique in type) without
> having a key of their choosing to select it?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3275/files#r22714410